src/config/backends/splash-image.ts

```diff
@@ -0,0 +1,196 @@
+import * as Bluebird from 'bluebird';
+import * as _ from 'lodash';
+import { fs } from 'mz';
+import * as path from 'path';
+
+import * as constants from '../../lib/constants';
+import log from '../../lib/supervisor-console';
+import {
+	bootMountPoint,
+	ConfigBackend,
+	ConfigOptions,
+	remountAndWriteAtomic,
+} from './backend';
+
+export class SplashImage extends ConfigBackend {
+	private static readonly BASEPATH = path.join(bootMountPoint, 'splash');
+	private static readonly DEFAULT = path.join(
+		SplashImage.BASEPATH,
+		'balena-logo-default.png',
+	);
+	private static readonly FILENAMES = ['balena-logo.png', 'resin-logo.png'];
+	private static readonly PREFIX = `${constants.hostConfigVarPrefix}SPLASH_`;
+	private static readonly CONFIGS = ['image'];
+	private static readonly DATA_URI_REGEX = /^data:(.+);base64,(.*)$/;
+
+	// Check the first 8 bytes of a buffer for a PNG header
+	// Source: https://github.com/sindresorhus/is-png/blob/master/index.js
+	private isPng(buffer: Buffer) {
+		if (!buffer || buffer.length < 8) {
+			return false;
+		}
+
+		return (
+			buffer[0] === 0x89 &&
+			buffer[1] === 0x50 &&
+			buffer[2] === 0x4e &&
+			buffer[3] === 0x47 &&
+			buffer[4] === 0x0d &&
+			buffer[5] === 0x0a &&
+			buffer[6] === 0x1a &&
+			buffer[7] === 0x0a
+		);
+	}
+
+	// Get the file path for the splash image for the underlying
+	// system
+	private async getSplashPath(): Promise<string> {
+		const [
+			// If no logo is defined, assume the file is `balena-logo.png`
+			splashFile = path.join(SplashImage.BASEPATH, 'balena-logo.png'),
+			// Read the splash dir (will throw if the path does not exist)
+		] = await Bluebird.resolve(fs.readdir(SplashImage.BASEPATH))
```

Comment:
This function to me seems like it is basically just returning balena-logo.png unless resin-logo.png exists. This is because the splashFile variable defaults to balena-logo.png and then loads the files in the directory. These files are filtered against the FILENAMES array which is balena-logo.png and resin-logo.png. 

If balena-logo.png exists then splashFile was already set to that so it is now set to the same value. If the files array contains resin-logo.png then that is set to the splashFile. If the directory contains both then splashFile is whichever file is at index 0 which I'd guess is alphabetical order (I just did ls -l and my system lists in alphabetical order).

If this is correct then it should just check if resin-logo.png exists else return balena-logo.png to simplify it.

Reply:
No, that is not correct. The function looks at the directory and checks that either of the files in `FILENAMES` exist
- If both exist, it will return the first (which might be an error, since it might be implementation or OS dependent)
- If only one exists, it returns that file
- If none exist, assume is `balena-logo.png`.

A better way would probably be to check os version and always use `balena-logo` if >= 2.52 and `resin-logo` if < 2.52

Reply:
To me that sounds just like if resin-logo.png exists return that else return balena-.logo.png.

Reply:
no, because if both resin-logo and balena-logo exist you should default to balena-logo. I agree is confusing though

Reply:
I changed my mind on this. Checking os versions goes against the idea of the supervisor not being tied to balenaOS. Probably the best option would be to specify the splash image as a variable that can be modified on the container start. But probably not going to happen at this time. 

I am modifying the code so the result of readdir is sorted, to avoid any implementation traps.

Reply:
Oh I understand the distinction now...this wants to use balena-logo by default AND choose that over resin-logo if they both exist. However, the mechanism for choosing balena-logo over resin-logo isn't predicable or conveyed in the code well.

If you add the sort then that does fix that though!